api/tests/mocha/services/generate-xform.spec.js

```diff
@@ -45,13 +45,17 @@ describe('generate-xform service', () => {
       });
     };
 
-    const runTest = (dirname, err) => {
+    const runTest = (dirname, err, writeErr) => {
```

Comment:
Feels like we had to hack this function a bit, I usually don't support this type of functions in unit tests for many reasons, readability of test cases are compromised, it can get more complex over time, it usually feels that we need to unit test the util function for testing 😅

1. Better to not use it so every test case have full control of what's going on and don't get side effects
2. Or restructure this function, few ideas:
Because we have so many stubs inside this function and we might need them to behave differently, then we can pass the stubs as map object in the parameters:
```js
it ('should test something...', () => {
    // here we create our stubs as we need in this test case
   const stubs {
      stdout: sinon.stub().returns(something),
      stderr: sinon.stub.returns(something),
      stdin: {
          write: sinon.stub.throws(something)
       }
   };
   .... then pass the stubs to the util function
   runTest(param, stubs);
});

// util function takes stubs from parameter map object and if something doesn't some then default 
const runTest = (dirname, stubs) => {
  const spawned = {
        stdout: { on: stubs.stdout || sinon.stub() },
        stderr: { on: stubs.stderr || sinon.stub() },
        stdin: {
          setEncoding: stubs.setEncoding || sinon.stub(),
          write: stubs.write || sinon.stub(),
          write: stubs.write || writeStub,
          end: stubs.end || sinon.stub()
        },
        on: stubs.on || sinon.stub()

```
This will give back more control to the test cases. And will reduce complexity in the following lines bellow

Reply:
The problem is that `runTest` is called 6 times with almost the same large set of stubs each time, so repeating each time the dataset makes it kind of DRY anti-pattern,  but I got your point. I'll split out the code that creates the dataset from the rest of the code.